hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java

```diff
@@ -62,6 +63,19 @@ public void assignWatermarkKeys() {
         visitor.go(root);
     }
 
+    /**
+     *  Special method to extract input's watermark key for {@link SlidingWindowAggregatePhysicalRel}.
+     *  We need it, because {@link SlidingWindowP} operates on input watermark key, which may not be
+     *  calculated in {@link WatermarkKeysAssigner} due to nature of its algorithm : it calculates how
+     *  watermarks are propagated from scans to root relations. Sometimes  {@link SlidingWindowAggregatePhysicalRel}
+     *  may break watermark key propagation chain, but we still need a watermark key for {@link SlidingWindowP}.
+     */
+    public byte getInputWatermarkKey(SlidingWindowAggregatePhysicalRel swaRel) {
+        Map<Integer, MutableByte> inputWmMap = visitor.getRelToWmKeyMapping().get(swaRel.getInput());
+        assert !inputWmMap.isEmpty() : "Input rel for SlidingWindowAggregate must contain watermarked field";
+        return inputWmMap.values().iterator().next().getValue();
```

Comment:
is it enough to use first input watermark? what if the input has more watermarks (eg. it is a results of s2s join)?

Reply:
https://github.com/hazelcast/hazelcast/pull/24129/commits/11cd12095f407505d77c384af673401e282d6277